File: src/core/objects/commit/commit-person.ts
/**
 * Commit Person Structure:
 * ┌─────────────────────────────────────────────────────────────────┐
 * │ Name <email> timestamp timezone                                 │
 * └─────────────────────────────────────────────────────────────────┘
 */
export class CommitPerson {
  readonly name: string;
  readonly email: string;
  readonly timestamp: number;
  readonly timezone: string;

  constructor(name: string, email: string, timestamp: number, timezone: string) {
    this.name = this.validateName(name);
    this.email = this.validateEmail(email);
    this.timestamp = timestamp;
    this.timezone = timezone;
  }

  /**
   * Validates the name and throws an error if it is null or empty
   */
  private validateName(name: string): string {
    if (name == null || name.trim().length === 0) {
      throw new Error('Name cannot be null or empty');
    }
    return name.trim();
  }

  /**
   * Validates the email and throws an error if it is null or empty or does not contain an @
   */
  private validateEmail(email: string): string {
    if (email == null || email.trim().length === 0) {
      throw new Error('Email cannot be null or empty');
    }

    if (!email.includes('@')) {
      throw new Error(`Invalid email format: ${email}`);
    }
    return email.trim();
  }

  /**
   * Formats person information in Git's standard format:
   * "Name <email> timestamp timezone"
   */
  formatForGit(): string {
    const epochSeconds = this.timestamp.toString();
    const tzString = this.formatTimezone(this.timezone);
    return `${this.name} <${this.email}> ${epochSeconds} ${tzString}`;
  }

  /**
   * Formats the timezone in the format "+HHMM" or "-HHMM"
   */
  private formatTimezone(timezone: string): string {
    const totalSeconds = parseInt(timezone, 10);
    const hours = Math.floor(Math.abs(totalSeconds) / 3600);
    const minutes = Math.floor((Math.abs(totalSeconds) % 3600) / 60);
    const sign = totalSeconds >= 0 ? '+' : '-';
    return `${sign}${hours.toString().padStart(2, '0')}${minutes.toString().padStart(2, '0')}`;
  }

  /**
   * Parses person information from Git's format.
   */
  static parseFromGit(gitFormat: string): CommitPerson {
    // Pattern: "Name <email> timestamp timezone"
    const pattern = new RegExp('^(.+) <([^>]+)> (\\d+) ([+-]\\d{4})$');
    const matcher = pattern.exec(gitFormat);
    if (!matcher) {
      throw new Error(`Invalid person format: ${gitFormat}`);
    }

    const name = matcher[1] ?? '';
    const email = matcher[2] ?? '';
    const epochSeconds = parseInt(matcher[3] ?? '0', 10);
    const tzString = matcher[4] ?? '';
    const timezone = CommitPerson.parseTimezone(tzString);

    return new CommitPerson(name, email, epochSeconds, timezone);
  }

  /**
   * Parses the timezone from the string and returns the total offset in seconds
   */
  private static parseTimezone(tzString: string): string {
    if (tzString.length !== 5 || (tzString.charAt(0) !== '+' && tzString.charAt(0) !== '-')) {
      throw new Error(`Invalid timezone format: ${tzString}`);
    }

    const sign = tzString.charAt(0) === '+' ? 1 : -1;
    const hours = parseInt(tzString.substring(1, 3), 10);
    const minutes = parseInt(tzString.substring(3, 5), 10);

    // Return total offset in seconds, not formatted string
    const totalSeconds = sign * (hours * 3600 + minutes * 60);
    return totalSeconds.toString();
  }
}

File: sourcecontrol/pkg/objects/commit/commit_person.go
package commit

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// CommitPerson represents author or committer information in a Git commit.
//
// Commit Person Structure:
// ┌─────────────────────────────────────────────────────────────────┐
// │ Name <email> timestamp timezone                                 │
// └─────────────────────────────────────────────────────────────────┘
//
// Example: "John Doe <john@example.com> 1609459200 +0000"
type CommitPerson struct {
	Name  string
	Email string
	When  time.Time
}

// personPattern is the regex pattern for parsing Git person format
// Pattern: "Name <email> timestamp timezone"
var personPattern = regexp.MustCompile(`^(.+) <([^>]+)> (\d+) ([+-]\d{4})$`)

// NewCommitPerson creates a new CommitPerson with validation
func NewCommitPerson(name, email string, when time.Time) (*CommitPerson, error) {
	if err := validateName(name); err != nil {
		return nil, err
	}

	if err := validateEmail(email); err != nil {
		return nil, err
	}

	return &CommitPerson{
		Name:  strings.TrimSpace(name),
		Email: strings.TrimSpace(email),
		When:  when,
	}, nil
}

// FormatForGit formats person information in Git's standard format:
// "Name <email> timestamp timezone"
func (p *CommitPerson) FormatForGit() string {
	timestamp := p.When.Unix()
	_, offset := p.When.Zone()

	// Format timezone as +HHMM or -HHMM
	sign := "+"
	if offset < 0 {
		sign = "-"
		offset = -offset
	}
	hours := offset / 3600
	minutes := (offset % 3600) / 60

	return fmt.Sprintf("%s <%s> %d %s%02d%02d",
		p.Name, p.Email, timestamp, sign, hours, minutes)
}

// ParseCommitPerson parses person information from Git's format.
// Format: "Name <email> timestamp timezone"
// Example: "John Doe <john@example.com> 1609459200 +0000"
func ParseCommitPerson(gitFormat string) (*CommitPerson, error) {
	matches := personPattern.FindStringSubmatch(gitFormat)
	if matches == nil {
		return nil, fmt.Errorf("invalid person format: %s", gitFormat)
	}

	name := matches[1]
	email := matches[2]
	timestampStr := matches[3]
	timezoneStr := matches[4]

	timestamp, err := strconv.ParseInt(timestampStr, 10, 64)
	if err != nil {
		return nil, fmt.Errorf("invalid timestamp: %w", err)
	}

	location, err := parseTimezone(timezoneStr)
	if err != nil {
		return nil, fmt.Errorf("invalid timezone: %w", err)
	}

	when := time.Unix(timestamp, 0).In(location)

	return NewCommitPerson(name, email, when)
}

// String returns a human-readable representation
func (p *CommitPerson) String() string {
	return fmt.Sprintf("%s <%s> at %s", p.Name, p.Email, p.When.Format(time.RFC3339))
}

// Equal compares two CommitPerson instances for equality
func (p *CommitPerson) Equal(other *CommitPerson) bool {
	if other == nil {
		return false
	}
	return p.Name == other.Name &&
		p.Email == other.Email &&
		p.When.Unix() == other.When.Unix()
}

// validateName validates the person name
func validateName(name string) error {
	trimmed := strings.TrimSpace(name)
	if trimmed == "" {
		return fmt.Errorf("name cannot be empty")
	}
	return nil
}

// validateEmail validates the email address
func validateEmail(email string) error {
	trimmed := strings.TrimSpace(email)
	if trimmed == "" {
		return fmt.Errorf("email cannot be empty")
	}
	if !strings.Contains(trimmed, "@") {
		return fmt.Errorf("invalid email format: %s", email)
	}
	return nil
}

// parseTimezone parses timezone string like "+0530" or "-0800" and returns a Location
func parseTimezone(tzString string) (*time.Location, error) {
	if len(tzString) != 5 {
		return nil, fmt.Errorf("invalid timezone length: %s", tzString)
	}

	sign := tzString[0]
	if sign != '+' && sign != '-' {
		return nil, fmt.Errorf("invalid timezone sign: %c", sign)
	}

	hours, err := strconv.Atoi(tzString[1:3])
	if err != nil {
		return nil, fmt.Errorf("invalid timezone hours: %w", err)
	}

	minutes, err := strconv.Atoi(tzString[3:5])
	if err != nil {
		return nil, fmt.Errorf("invalid timezone minutes: %w", err)
	}

	offsetSeconds := hours*3600 + minutes*60
	if sign == '-' {
		offsetSeconds = -offsetSeconds
	}

	return time.FixedZone(tzString, offsetSeconds), nil
}
